### ISSUE TITLE
`!is` is treated as `is`

### ISSUE THREAD

**Issue by asashour** [dart-lang/linter#2995]
My bad luck drove me to write something like:

```dart
void f(Object? a) {
  print(a !is num);
}
void main() {
  f(1);
}
```

so in fact `a !is num` is exactly the same as `a is num`.

Is there anything we should do to prevent primitive ones like me from doing so?

**Comment by srawlins** [comment]
I'm going to throw this over to the linter as it does not feel like a rule that the analyzer would include. Since its based on _un_ formatted code, and that formatting the code would lead the issue to not be reported.

I'll also note that I'm not in favor of writing static analysis checks specifically for unformatted code, as it duplicates work between the formatter and the linter.

**Comment by bwilkerson** [comment]
> I'm not in favor of writing static analysis checks specifically for unformatted code ...

I agree with the conclusion, though for a different reason. However, I don't believe that the check has anything to do with the way the code is formatted, so I don't think that's a good reason for it to not be a hint.

The question is: how often is a user likely to want to ensure that an exception is thrown if the left operand is `null` immediately prior to checking that the left operand has a given type? This doesn't seem to me to be a common pattern for developers to use, especially given that you can explicitly exclude `null` based on the type (though the semantics of the two approaches are different), but we could certainly measure to see how often it occurs in practice.

But, given that it isn't necessarily wrong to write this code such a check might still have false positives, so for that reason I agree that it's probably more appropriate for the check to be a lint than a hint.

**Comment by pq** [comment]
I agree w/ @bwilkerson and @lrhn that this would make a good lint (over a hint).

**Comment by srawlins** [comment]
> The question is: how often is a user likely to want to ensure that an exception is thrown if the left operand is null immediately prior to checking that the left operand has a given type?

That's a really good point!

**Comment by eggfly** [comment]
kotlin use !! to convert nullsable type to none-null type:
```
    fun f(a: Any?) {
        val result = a !! is String
        print(result)
    }
    f(1111)
```
when not space between !! and is there will be an compilation error:
```
    fun f(a: Any?) {
        val result = a !!is String
        print(result)
    }
    f(1111)
```

main.kt: (30, 24): Unexpected tokens (use ';' to separate expressions on the same line)

maybe this is not only a lint, but we should change dart frontend compiler?

**Comment by natebosch** [comment]
+1 on a diagnostic for this usage. If someone really does want to throw for a null value in the same expression as an `is` check, they can wrap with parens.
